Add object type for StationId

cbs228 opened the pull request:
Switch to a "smart object" representation for callsigns throughout all of ardopcf. This is like `Locator` (#93), but the impact of this PR is far greater since callsigns are much more frequent and critical than grid squares.

The new `StationId` object has the following advantages over the older processors:

* Full support for seven-character callsigns

* The **canonical representation** of callsigns is guaranteed. Station IDs are canonicalized via a round-trip through the compressor.
  
* Packages the following representations together:

  * Canonical full string (`N0CALL-A`)
  * Callsign string only (`N0CALL`)
  * SSID (`A`)
  * `Packed6` wireline representation

* Includes methods for printing and parsing arrays for `MYAUX` support.

* Fallible functions guarantee populated and valid StationId for all frames that require them.

* Catches errors that legacy code misses, including erroneously truncating `N0CALL-16` to `N0CALL-1`.

* Improved test coverage

The legacy (de)compressors are removed.

cbs228 committed:
station-id: add object type for StationId

We can make the callsign transmit process more robust by adding
an object type to represent the `StationId`. The conversions and
logic are much more formalized. Improvements include:

* Fallible functions that return error codes

* A more robust string parser that does not rely on the unbounded
  strcspn() function

* Packages human- and machine-readable formats together. Sized
  structs guarantee that the required amount of buffer
  space is always available.

* Canonicalizes input with a round-trip through the compressor.
  Two human-readable outputs are available

  1. canonical ARDOP string representation, like `N0CALL-A`
  2. individual callsign and ssid strings

* Catches errors that legacy code misses, including erroneously
  truncating `N0CALL-16` to `N0CALL-1`.

* Includes array processing functions for `MYAUX`

cbs228 commented:
7b53bc6839a263e808f1d884846ffce6c7391ad2: ardopc also includes a `CQ` callsign that doesn't appear anywhere in the spec.

pflarue reviewed (commented):
I reviewed only part of your pull request.

I think that the idea of using custom structs here is a good idea, though I added some questions about some of your other design choices.  Pretending that we are using an object oriented language, as you alluded to in your description of this pull request, is something I fully support.  If I thought we could do it in a reasonable amount of time and without breaking things, I'd be in favor of attempting a conversion from c to c++ so that we could take advantage of its object oriented structure as well as some of its inherently safer standard features.

While I agree that the changes in this pull request, and similar changes that I assume you will want to make continuing forward, are likely to be useful, I'm a bit frustrated by them and I'm unsure what to do about it.  So, I'll explain here how I am feeling.  

Providing the level of detailed review that I feel is appropriate takes a lot of time.  I'd rather spend that time working on visible improvements, trying to fix things that we know are causing problems, or implementing testing that may identify broken things we haven't yet noticed.  Fixing things that aren't broken is not high on my list of priorities while there are outstanding issues that I think deserve attention.  Some of the changes I have made recently don't necessarily fit into these categories, but they were changes whose effort seemed justifiable to improve the large scale maintainability of the project.

Part of the problem is that due to my own limitations, mentally compiling your work so that I fully understand it and can intelligently comment on it is a slow process.

One possible solution to this is for you to take over as primary developer of ardopcf.  I'd then be happy to submit pull requests to your project for the sort of changes that I think are important.  An alternative is that we agree to discuss the type of changes you plan to make in words or pseudo-code before you implement them.  Then, I limit my reviews of your pull requests to compiling and running tests (automated and manual) on the various hardware/OS combinations that we are trying to support.  This is contrary to my instincts, but it would probably allow us to each contribute to this project in the ways that will lead to the best rate of improvement.  I'm willing to consider these or another approach that you might suggest because I believe that your programming skills exceed my own, and because you seem genuinely interested in the success of this project.

[edit]  Improving the user documentation is another thing I'd be happy to spend my time on.

cbs228 commented:
> If I thought we could do it in a reasonable amount of time and without breaking things, I'd be in favor of attempting a conversion from c to c++

In my opinion, C++ just wraps every utter failure of C in a shiny chromium coating. Don't get me wrong, there are a *lot* of nice-to-haves… but if I was going to tackle a new-language rewrite, I'd rewrite it in Rust 🦀. It is a much more pleasant and robust language.

One shortcoming of Rust is that the build tools are very heavyweight. Rust has also some amazing reusable modules, but those modules will also add tons of transitive dependencies very quickly.

> Providing the level of detailed review that I feel is appropriate takes a lot of time. I'd rather spend that time working on visible improvements,

Yeah, I do a *lot* of code review. It's a tough skill to master. It's frequently no fun.

> One possible solution to this is for you to take over as primary developer of ardopcf.

I was kind of hoping to avoid that. I maintain so many things already that keeping up with them all is a challenge. I'm not much of a researcher, either.

* I have some DSP knowledge but very little experience with HF-specific challenges like severe multipath and fading. Modem design is a bit beyond me.

* I also don't get a lot of RF air time. My station is QRP portable and usually not assembled. The QRM is about as bad as it gets here. I've got maybe 8 watts continuous. I can reach the low-rate PSK modes, sometimes, but a QAM uplink is well out of my reach. Unless I make a substantial investment in upgrades, most of the modes aren't all that useful to me.

I have been tempted many times to embark on my own rewrite of ARDOP, but I don't think I really have anything new to bring to this scene. Incrementally better, perhaps, but not new.

> This is contrary to my instincts

I would encourage you to mark major changes in maintainership with something that requires user-facing action: for example, to change a project URL or package name. That prevents the changeover from passing silently, which is a serious security risk.

It's probably going to take a me a few days to get to the revisions, but it shall be done.

pflarue commented:
While I'm getting back used to working with C, and I think I could do the same for C++, everything I've read about Rust suggests a steep learning curve that I'm not interested in climbing at this time.

A good feature of Ardop for QRP use is that while the QRP station may only be able to reliably transmit using the slower more robust frame types, if the other station is using more power, the QRP station may receive data transmitted via faster frame types.  Of course, with both QRP and QRM, it is difficult.

Because we are constrained by the Ardop specification and not breaking compatibility with other Ardop implementations, incremental improvements are all we can hope for.  However, I believe that the combination of many of those incremental improvements can noticably improve reliability, performance, and ease of use.

Any change in maintainer would mean that the new master repository would be under somebody else's GitHub acct.  If somebody who I thought had the skills and interest in taking over wanted to do so, I'd then add a note to this repository directing people to that one.  However, it doesn't look like that is happening.  So, I'll continue.

Take your time.  As I wrote, I've got plenty of other things to focus on.  Right now I'm enhancing some diagnostic features that are helping me to better understand Issue #78.

cbs228 reviewed (commented):
> everything I've read about Rust suggests a steep learning curve

Rust is super hard until you "get it." When you "get it," you discover that Rust code can't look like C, Python, or Java, and that certain design patterns are necessary to reduce mutability and to control lifetimes. The two hardest things IMO are:

* Lifetime specifiers, which are less necessary these days but still crop up
* Generics (templates), and getting generic trait bounds correct. But that's hard in any language.

But by far the biggest disadvantage is having to start over from scratch. Re-using C in in Rust requires a lot of very messy interfacing and dealing with lots of build environments.

Anyway, a4029c0c38888ddb469240bd0cfa13a933c0a656 gets the easy stuff out of the way. There are a few threads that need more discussion.

I'm getting closer to having a working `Locator` object to store grid squares, but I'll omit that until we get the existing objects the way you want them.

pflarue reviewed (commented):
Additional comments, questions, discussion are welcome, but please do not submit additional changes until this PR is ready to merge.  I don't want to spend much more time on this.

cbs228 commented:
> please do not submit additional changes until this PR is ready to merge

I have the minimal fixes for the issues you've reported, but I assume from your comment that you don't want to see them until this is fully integrated in its final form. Correct?

> What do you propose to do with the error result?

* **Outgoing frames**: all errors will be prevented at the source. If the host attempts to use `MYCALL` or `GRIDSQUARE` to set an invalid callsign or grid square, the TNC will reject the attempt with the appropriate `FAULT` code. Frames with invalid callsigns will not be transmitted.

* **Incoming frames**: a frame with an invalid callsign should be considered invalid, discarded with a logged warning, and not acted upon any further. ID frames with invalid callsigns won't be reported as `IDF` on the host interface. ARQ calls with an invalid source callsign will elicit no reply. An invalid grid square will *not* invalidate the frame but will also not be reported to the host.

Basically, my intent is to treat these errors like they would be handled in IPv4: bad data → drop frame. Given the very wide acceptable criteria for a callsign, I do not think we will regret the breakage that this deliberately introduces.

> Your response to a similar question that I wrote earlier suggests that perhaps these return codes are primarily intended for testing purposes, and will thus be ignored during normal operation.

Correct, the calling functions will simply check if the error code is non-zero. Future revisions will add logging statements at the source of the error. That way, the calling function won't need to `switch` on it.

I'm going to see what the best solution to #71 will be before I come back to this.

pflarue commented:
Yes, please continue until you have integrated the changes.

Regarding bad callsigns, your plan sounds good.  For from incoming frames, you may (or may not) have noticed a comment I added a while back in Decode4FSKConReq().  While I was updating the RS correction handling I noticed that these frames do include a few RS correction bytes, but do not use a CRC that can be used to verify whether or not the corrected results are actually correct.  So, I added a comment wondering whether the encoded callsigns should be checked for validity as a way to discard some incorrectly decoded ConReq frames.  This change you are working on should do that, which is good.

cbs228 committed:
soundinput: don't write to intReceivedLeaderLen

When a ConAck frame is received, the frame decoder writes the
leader timing to the ARQ.c module. But ARQ.c also reads the
timing from SoundInput.c [1]. The write within SoundInput is
completely unnecessary. Remove it.

[1]: src/common/ARQ.c:1880
     `case ISS: ... if (ARQState == ISSConReq)`

cbs228 committed:
common: widen CALL_BUF_SIZE to 11 bytes

To properly support 7-character callsigns, align ARDOP's legacy
string callsign representation with the new StationId object.

The callsign buffer within `GenerateSessionID()` was potentially
too short to accommodate two, full-length callsigns. Increase the
buffer size. It is not known if the increased buffer length is
compatible with other ARDOP implementations.

cbs228 committed:
common: replace host interface callsign globals

Replace the following host interface globals with `StationId`:

* `Callsign`: stores the local station callsign (`MYCALL`)

* `ConnectToCall`: holds destination callsign for
  "`ARQCALL`" and "`PING`"

* AuxCalls: array of alternative callsigns. Its companion,
  AuxCallsLength is now a size_t.

This refactor deliberately avoids touching the ARQ logic, which has
several other callsign-like globals. To smooth over the API
mismatch, we add some temporary local `StationId` to convert from
string. These will be removed by later patches.

cbs228 committed:
hostinterface: add parser for station ID + attempt count

The `ARQCALL` and `PING` commands both accept a target station ID
and an attempt count, like

    N0CALL-A 5

Add a parser specifically for these types of string arguments.

cbs228 committed:
soundinput: replace callsign strings with smart objects

Replace all string-based callsigns in `SoundInput.c` with the new
`StationId` smart object. The ConReq, Ping, and IDFrame decoders
now take and populate `StationId` instead of string buffers.

Both the ConReq and Ping decoders communicate the decode to
higher-level logic (i.e., ARQ) via string buffers of the form
"`CALLER TARGET`." While we continue to populate this string,
we will eventually replace this with the new globals:

* `LastDecodedStationCaller`
* `LastDecodedStationTarget`

which hold the last successfully-decoded station IDs. We take
care to update these consistently and to zero them out with each
new frame.

The `lastGoodID` string appears to exist only to feed a UI
element. It is refactored out.

cbs228 committed:
arq: replace callsign globals

Replace text-based callsign globals within the ARQ logic:

The ConReq and Ping frame decoders now pass decoded callsigns to
to the ARQ logic via the following globals instead of the
`bytData` frame data global:

* `LastDecodedStationCaller`
* `LastDecodedStationTarget`

String buffers in the ARQ session logic are replaced:

* `strRemoteCallsign` → `ARQStationRemote`
* `strLocalCallsign` → `ARQStationLocal`
* `strFinalIDCallsign` → `ARQStationFinalId`

The existing logic is changed as little as possible.

cbs228 committed:
common: drop legacy ASCII (de)compressors

The legacy callsign compressor and decompressor functions have
served their purpose. With the migration to `StationId`, they are
no longer needed. Remove.

cbs228 committed:
common: CALL_BUF_SIZE → STATIONID_BUF_SIZE

Eliminate the duplicate `CALL_BUF_SIZE` definition, which is no
longer necessary. Replace with `STATIONID_BUF_SIZE` from
`StationId.h`.

Since the StationId buffer size is defined in terms of this macro,
this change enforces a consistent definition everywhere in
ardopcf.

cbs228 commented:
2d7a1731197b8717343b1a322a299354da2e343f should fully integrate the new `StationId` type and remove the old (de)compressors. I have done basic loopback testing of this on Linux. ARQ sessions, pings, and ID frames appear to be working. I have tested connections between both 2d7a173-to-2d7a173 and 2d7a173-to-develop.

This version of `StationId` is a little different than in the previous review, but the basic concept is the same. Significant changes include:

* A new `str` field that contains the full callsign-SSID pre-formatted as a C string. This is used in too many places for us not to have it.

* New array-processing functions for MYAUX.

* Logging of wireline bytes that fail decode

* Prefer copying Packed6 to buffer with `stationid_to_buffer()`, which checks first that the StationId is populated and valid.

The basic idea remains the same.

This PR is broken out into several individual commits. The individual commits gradually roll out the new `StationId` to all parts of ardopcf. If you encounter a regression bug anywhere, I would appreciate a `git-bisect` to identify which commit is responsible.
